This will review Module 1 - 4

The first section combines module 1&2

In [19]:
import pandas as pd

In [20]:
pokemon = pd.read_csv('data/pokemon.csv')
pokemon

name  deck_no  attack  defense  sp_attack  sp_defense  speed  \
0     Bulbasaur        1      49       49         65          65     45   
1       Ivysaur        2      62       63         80          80     60   
2      Venusaur        3     100      123        122         120     80   
3    Charmander        4      52       43         60          50     65   
4    Charmeleon        5      64       58         80          65     80   
..          ...      ...     ...      ...        ...         ...    ...   
796  Celesteela      797     101      103        107         101     61   
797     Kartana      798     181      131         59          31    109   
798    Guzzlord      799     101       53         97          53     43   
799    Necrozma      800     107      101        127          89     79   
800    Magearna      801      95      115        130         115     65   

     capture_rt  total_bs     type  gen  legendary  
0            45       318    grass    1          0  
1            45       405    grass    1          0  
2            45       625    grass    1          0  
3            45       309     fire    1          0  
4            45       405     fire    1          0  
..          ...       ...      ...  ...        ...  
796          25       570    steel    7          1  
797         255       570    grass    7          1  
798          15       570     dark    7          1  
799           3       600  psychic    7          1  
800           3       600    steel    7          1  

[801 rows x 12 columns]

In [24]:
pokemon.columns

Index(['pokemon', 'deck_no', 'attack', 'defense', 'sp_attack', 'sp_defense',
       'speed', 'capture_rate', 'total_bs', 'type', 'gen', 'legendary'],
      dtype='object')

In [25]:
pokemon.shape

(801, 12)

In [26]:
pokemon.head(3)

pokemon  deck_no  attack  defense  sp_attack  sp_defense  speed  \
0  Bulbasaur        1      49       49         65          65     45   
1    Ivysaur        2      62       63         80          80     60   
2   Venusaur        3     100      123        122         120     80   

   capture_rate  total_bs   type  gen  legendary  
0            45       318  grass    1          0  
1            45       405  grass    1          0  
2            45       625  grass    1          0

How to read different data sets.

pd.read_csv('https://WEBPAGE.CSV)
pd.read_csv('textdoc.txt', delimiter='/T')
pd.read_excel('Excel.xls', sheet_name='sheet1')

Arguments for pd.read_csv() and pd.read_excel()

index_col - indicates which column will be the index label
dataset = pd.read_csv('dataset.csv', index_col= 'column name')

header - used to indicate what row the data starts on
dataset = pd.read_csv('dataset.csv', header=2)

nrows - only lows the number of rows you want to view
dataset = pd.read_csv('dataset.csv', nrows=7)

usecols - selects only the columns you choose 
dataset = pd.read_csv('dataset.csv', usecols=[0,9,4]

Renaming Columns 

can rename multiple columns using comma between each column
.rename()

In [21]:
pokemon=pokemon.rename(columns={'name':'pokemon', 'capture_rt':'capture_rate'})
pokemon

pokemon  deck_no  attack  defense  sp_attack  sp_defense  speed  \
0     Bulbasaur        1      49       49         65          65     45   
1       Ivysaur        2      62       63         80          80     60   
2      Venusaur        3     100      123        122         120     80   
3    Charmander        4      52       43         60          50     65   
4    Charmeleon        5      64       58         80          65     80   
..          ...      ...     ...      ...        ...         ...    ...   
796  Celesteela      797     101      103        107         101     61   
797     Kartana      798     181      131         59          31    109   
798    Guzzlord      799     101       53         97          53     43   
799    Necrozma      800     107      101        127          89     79   
800    Magearna      801      95      115        130         115     65   

     capture_rate  total_bs     type  gen  legendary  
0              45       318    grass    1          0  
1              45       405    grass    1          0  
2              45       625    grass    1          0  
3              45       309     fire    1          0  
4              45       405     fire    1          0  
..            ...       ...      ...  ...        ...  
796            25       570    steel    7          1  
797           255       570    grass    7          1  
798            15       570     dark    7          1  
799             3       600  psychic    7          1  
800             3       600    steel    7          1  

[801 rows x 12 columns]

Dropping Columns
.drop(columns=['column1','column2'])

In [23]:
pokemondrop=pokemon.drop(columns=['legendary','deck_no'])
pokemondrop

pokemon  attack  defense  sp_attack  sp_defense  speed  capture_rate  \
0     Bulbasaur      49       49         65          65     45            45   
1       Ivysaur      62       63         80          80     60            45   
2      Venusaur     100      123        122         120     80            45   
3    Charmander      52       43         60          50     65            45   
4    Charmeleon      64       58         80          65     80            45   
..          ...     ...      ...        ...         ...    ...           ...   
796  Celesteela     101      103        107         101     61            25   
797     Kartana     181      131         59          31    109           255   
798    Guzzlord     101       53         97          53     43            15   
799    Necrozma     107      101        127          89     79             3   
800    Magearna      95      115        130         115     65             3   

     total_bs     type  gen  
0         318    grass    1  
1         405    grass    1  
2         625    grass    1  
3         309     fire    1  
4         405     fire    1  
..        ...      ...  ...  
796       570    steel    7  
797       570    grass    7  
798       570     dark    7  
799       600  psychic    7  
800       600    steel    7  

[801 rows x 10 columns]

.loc 
dataset.loc[5:10] - obtains rows 5:10

dataset.loc['column name start':'column name end'] - gives columns for that frame

dataset.loc[:,"column name"] gives all rows for that columns name

dataset.loc[2:5,"column name"] gives rows 2-5 for that columns name

dataset.loc[[13,5,9], ['columnname','columnname2']]

you can reorganize the data frame using the last example listed:

In [38]:
pokemon_cut=pokemon.loc[[13,5,9], ['pokemon','type','total_bs']]
pokemon_cut

pokemon  type  total_bs
13     Kakuna   bug       205
5   Charizard  fire       634
9    Caterpie   bug       195

In [42]:
pokemon_sort=pokemon_cut.sort_values('total_bs')
pokemon_sort

pokemon  type  total_bs
9    Caterpie   bug       195
13     Kakuna   bug       205
5   Charizard  fire       634

to sort from highest to lowest:
You need to remember because False is boolean it's first letter is capatalized.

In [43]:
pokemon_sort_highest=pokemon_sort.sort_values('total_bs', ascending=False)
pokemon_sort_highest

pokemon  type  total_bs
5   Charizard  fire       634
13     Kakuna   bug       205
9    Caterpie   bug       195

In [44]:
pokemon_stats=pokemon.describe()
pokemon_stats

deck_no      attack     defense   sp_attack  sp_defense       speed  \
count  801.000000  801.000000  801.000000  801.000000  801.000000  801.000000   
mean   401.000000   77.857678   73.008739   71.305868   70.898876   66.184769   
std    231.373075   32.158820   30.769159   32.353826   27.929483   28.935805   
min      1.000000    5.000000    5.000000   10.000000   20.000000    5.000000   
25%    201.000000   55.000000   50.000000   45.000000   50.000000   45.000000   
50%    401.000000   75.000000   70.000000   65.000000   65.000000   64.000000   
75%    601.000000  100.000000   90.000000   91.000000   90.000000   85.000000   
max    801.000000  185.000000  230.000000  194.000000  230.000000  180.000000   

       capture_rate    total_bs         gen   legendary  
count    801.000000  801.000000  801.000000  801.000000  
mean      99.118602  428.377029    3.690387    0.087391  
std       76.334950  119.203577    1.930420    0.282583  
min        3.000000  180.000000    1.000000    0.000000  
25%       45.000000  320.000000    2.000000    0.000000  
50%       60.000000  435.000000    4.000000    0.000000  
75%      170.000000  505.000000    5.000000    0.000000  
max      255.000000  780.000000    7.000000    1.000000

In [45]:
pokemon_stats_all=pokemon.describe(include='all')
pokemon_stats_all

pokemon     deck_no      attack     defense   sp_attack  sp_defense  \
count       801  801.000000  801.000000  801.000000  801.000000  801.000000   
unique      800         NaN         NaN         NaN         NaN         NaN   
top     Porygon         NaN         NaN         NaN         NaN         NaN   
freq          2         NaN         NaN         NaN         NaN         NaN   
mean        NaN  401.000000   77.857678   73.008739   71.305868   70.898876   
std         NaN  231.373075   32.158820   30.769159   32.353826   27.929483   
min         NaN    1.000000    5.000000    5.000000   10.000000   20.000000   
25%         NaN  201.000000   55.000000   50.000000   45.000000   50.000000   
50%         NaN  401.000000   75.000000   70.000000   65.000000   65.000000   
75%         NaN  601.000000  100.000000   90.000000   91.000000   90.000000   
max         NaN  801.000000  185.000000  230.000000  194.000000  230.000000   

             speed  capture_rate    total_bs   type         gen   legendary  
count   801.000000    801.000000  801.000000    801  801.000000  801.000000  
unique         NaN           NaN         NaN     18         NaN         NaN  
top            NaN           NaN         NaN  water         NaN         NaN  
freq           NaN           NaN         NaN    114         NaN         NaN  
mean     66.184769     99.118602  428.377029    NaN    3.690387    0.087391  
std      28.935805     76.334950  119.203577    NaN    1.930420    0.282583  
min       5.000000      3.000000  180.000000    NaN    1.000000    0.000000  
25%      45.000000     45.000000  320.000000    NaN    2.000000    0.000000  
50%      64.000000     60.000000  435.000000    NaN    4.000000    0.000000  
75%      85.000000    170.000000  505.000000    NaN    5.000000    0.000000  
max     180.000000    255.000000  780.000000    NaN    7.000000    1.000000

all of the describe stats can be asked for individually

In [46]:
pokemon_one=pokemon[['attack']]
pokemon_one

attack
0        49
1        62
2       100
3        52
4        64
..      ...
796     101
797     181
798     101
799     107
800      95

[801 rows x 1 columns]

In [48]:
pokemon_one_mean=pokemon_one.mean()
pokemon_one_mean

attack    77.857678
dtype: float64

In [49]:
pokemon_type=pokemon[['type']]
pokemon_type

type
0      grass
1      grass
2      grass
3       fire
4       fire
..       ...
796    steel
797    grass
798     dark
799  psychic
800    steel

[801 rows x 1 columns]

In [52]:
pokemon_type_count=pokemon_type.value_counts()
pokemon_type_count

type    
water       114
normal      105
grass        78
bug          72
psychic      53
fire         52
rock         45
electric     39
ground       32
poison       32
dark         29
fighting     28
ghost        27
dragon       27
steel        24
ice          23
fairy        18
flying        3
dtype: int64

In [59]:
import altair as alt

chart = alt.Chart(pokemon).mark_bar().encode(
    x='type',
    y='capture_rate'
)
chart

alt.Chart(...)

In [60]:
plot_df = pd.DataFrame(pokemon.rename(columns={'capture_rt':'capture_rate'})
.assign(AD_total=pokemon['defense']+ pokemon['attack']))
pokemon_plot = alt.Chart(plot_df, width=500, height=300).mark_circle().encode(
    x='AD_total',
    y='capture_rate'
)
pokemon_plot

alt.Chart(...)

In [66]:
pokemon_plot2=alt.Chart(plot_df, width=500, height=300).mark_circle(color='green', size=80, opacity=0.3).encode(
    x='AD_total',
    y='capture_rate'
).properties(title='Capture Rate and Attack/Defence Total')
pokemon_plot2

alt.Chart(...)

If the data isn't running properlly you set the data type in Altair
.encode(
    x=alt.X('Ad_total:N')
    y=alt.Y('capture_rate:O')
)
Ordinal= O - descrete or ordered data
Nominal= N - discrete unordered data
Quantitative = Q - continious quantity
Temporal = T - time or date value

Below is terrible example - but it does outline the way that the code is formatted.

This sets the amount of bars allowed in the chart:

encode(
    x=alt.X('Ad_total:N', bin=alt.bin(maxbins=20))
    y=alt.Y('capture_rate:O')
    )




In [76]:
pokemon_plot3=alt.Chart(plot_df, width=500, height=300).mark_circle(color='green', size=80, opacity=0.3).encode(
    x=alt.X('Ad_total:N'),
    y=alt.Y('capture_rate:Q')
).properties(title='Capture Rate and Attack/Defence Total')
pokemon_plot3

alt.Chart(...)

.agg and grouping

In [77]:
pokemon_groupby = pokemon.groupby('type')
pokemon_groupby

If you see below it groups all the rows that belong to each of the select columns

In [78]:
pokemon_groupby.groups

{'bug': [9, 10, 11, 12, 13, 14, 45, 46, 47, 48, 122, 126, 164, 165, 166, 167, 192, 203, 204, 211, 212, 213, 264, 265, 266, 267, 268, 282, 283, 289, 290, 291, 312, 313, 400, 401, 411, 412, 413, 414, 415, 468, 539, 540, 541, 542, 543, 544, 556, 557, 587, 588, 594, 595, 615, 616, 631, 635, 636, 648, 663, 664, 665, 735, 736, 737, 741, 742, 766, 767, 793, 794], 'dark': [196, 197, 214, 227, 228, 260, 261, 301, 358, 429, 460, 490, 508, 509, 558, 559, 569, 570, 623, 624, 628, 629, 632, 633, 634, 685, 686, 716, 798], 'dragon': [146, 147, 148, 333, 370, 371, 372, 379, 380, 383, 442, 443, 444, 609, 610, 611, 620, 642, 643, 645, 703, 704, 705, 717, 781, 782, 783], 'electric': [24, 25, 80, 81, 99, 100, 124, 134, 144, 171, 178, 179, 180, 238, 242, 308, 309, 310, 311, 402, 403, 404, 416, 461, 465, 478, 521, 522, 586, 601, 602, 603, 641, 693, 694, 701, 776, 784, 795], 'fairy': [34, 35, 172, 174, 175, 208, 209, 467, 668, 669, 670, 681, 682, 683, 684, 699, 715, 763], 'fighting': [55, 56, 65, 66, 67, 105

In [79]:
pokemon_groups_mean = pokemon_groupby.mean()
pokemon_groups_mean

deck_no      attack     defense  sp_attack  sp_defense  \
type                                                                  
bug       395.916667   70.125000   70.847222  56.652778   62.513889   
dark      490.034483   87.793103   70.517241  74.517241   69.068966   
dragon    516.296296  106.407407   86.259259  89.592593   83.629630   
electric  380.102564   70.820513   61.820513  87.538462   70.051282   
fairy     467.000000   62.111111   68.166667  81.500000   87.777778   
fighting  404.035714   99.178571   66.392857  50.107143   63.428571   
fire      376.403846   81.500000   67.788462  87.730769   71.538462   
flying    690.000000   66.666667   65.000000  84.000000   70.000000   
ghost     498.814815   72.740741   79.518519  82.444444   78.296296   
grass     410.128205   73.769231   70.871795  74.320513   69.230769   
ground    371.000000   94.812500   83.906250  51.937500   62.281250   
ice       426.217391   73.304348   71.913043  77.434783   76.130435   
normal    367.504762   75.161905   59.695238  56.980952   63.342857   
poison    314.375000   72.656250   70.031250  61.562500   65.531250   
psychic   429.735849   65.566038   69.264151  92.603774   85.735849   
rock      427.266667   90.666667   96.266667  63.200000   73.377778   
steel     486.708333   93.083333  120.208333  72.708333   82.208333   
water     341.166667   73.307018   73.482456  74.061404   71.798246   

              speed  capture_rate    total_bs       gen  legendary  
type                                                                
bug       63.569444    119.833333  380.430556  3.763889   0.041667  
dark      75.310345     84.482759  449.758621  4.275862   0.103448  
dragon    73.888889     40.481481  522.777778  4.296296   0.259259  
electric  85.410256    106.974359  436.153846  3.461538   0.128205  
fairy     53.666667    116.944444  427.166667  4.277778   0.055556  
fighting  64.285714    103.750000  414.821429  3.750000   0.000000  
fire      73.346154     72.403846  450.634615  3.634615   0.096154  
flying    99.666667     79.333333  453.333333  5.666667   0.333333  
ghost     58.333333     93.814815  434.703704  4.370370   0.037037  
grass     59.025641    106.935897  412.576923  3.935897   0.051282  
ground    59.968750    108.312500  426.093750  3.343750   0.062500  
ice       62.739130     94.173913  433.608696  3.565217   0.086957  
normal    68.961905    120.647619  401.295238  3.495238   0.028571  
poison    64.187500    128.437500  399.562500  3.093750   0.000000  
psychic   75.150943     86.660377  461.264151  3.849057   0.320755  
rock      57.422222     75.866667  447.266667  3.800000   0.088889  
steel     56.583333     58.916667  491.583333  4.208333   0.250000  
water     63.921053     99.412281  426.789474  3.210526   0.052632

If you see above you can take the column and do calculations on the column for each of the other columns.

You can then select one of the groupby sections, and it will list all the of the rows that belong to that group.

In [82]:
pokemon_groupby.get_group('rock')

pokemon  deck_no  attack  defense  sp_attack  sp_defense  speed  \
73      Geodude       74      80      100         30          30     20   
74     Graveler       75      95      115         45          45     35   
75        Golem       76     120      130         55          65     45   
94         Onix       95      45      160         30          45     70   
137     Omanyte      138      40      100         90          55     35   
138     Omastar      139      60      125        115          70     55   
139      Kabuto      140      80       90         55          45     55   
140    Kabutops      141     115      105         65          70     80   
141  Aerodactyl      142     135       85         70          95    150   
184   Sudowoodo      185     100      115         30          65     30   
245    Larvitar      246      64       50         45          50     41   
246     Pupitar      247      84       70         65          70     51   
247   Tyranitar      248     164      150         95         120     71   
298    Nosepass      299      45      135         45          90     30   
336    Lunatone      337      55       65         95          85     70   
337     Solrock      338      95       85         55          65     70   
344      Lileep      345      41       77         61          87     23   
345     Cradily      346      81       97         81         107     43   
346     Anorith      347      95       50         40          50     75   
347     Armaldo      348     125      100         70          80     45   
376    Regirock      377     100      200         50         100     50   
407    Cranidos      408     125       40         30          30     58   
408   Rampardos      409     165       60         65          50     58   
409    Shieldon      410      42      118         42          88     30   
410   Bastiodon      411      52      168         47         138     30   
437      Bonsly      438      80       95         10          45     10   
475   Probopass      476      55      145         75         150     40   
523  Roggenrola      524      75       85         25          25     15   
524     Boldore      525     105      105         50          40     20   
525    Gigalith      526     135      130         60          80     25   
565      Archen      566     112       45         74          45     70   
566    Archeops      567     140       65        112          65    110   
638   Terrakion      639     129       90         72          90    108   
687     Binacle      688      52       67         39          56     50   
688  Barbaracle      689     105      115         54          86     68   
695      Tyrunt      696      89       77         45          45     48   
696   Tyrantrum      697     121      119         69          59     71   
697      Amaura      698      59       50         67          63     46   
698     Aurorus      699      77       72         99          92     58   
702     Carbink      703      50      150         50         150     50   
718     Diancie      719     160      110        160         110    110   
743    Rockruff      744      65       40         30          40     60   
744    Lycanroc      745     115       75         55          75     82   
773      Minior      774     100       60        100          60    120   
792    Nihilego      793      53       47        127         131    103   

     capture_rate  total_bs  type  gen  legendary  
73            255       300  rock    1          0  
74            120       390  rock    1          0  
75             45       495  rock    1          0  
94             45       385  rock    1          0  
137            45       355  rock    1          0  
138            45       495  rock    1          0  
139            45       355  rock    1          0  
140            45       495  rock    1          0  
141            45       615  rock    1          0  
184            65       410  rock    

you can use various other codes with grouby
.mean
.value_counts
.max
.min
.size

In [90]:
pokemon_group_other=pokemon_groupby.size()
pokemon_group_other

type
bug          72
dark         29
dragon       27
electric     39
fairy        18
fighting     28
fire         52
flying        3
ghost        27
grass        78
ground       32
ice          23
normal      105
poison       32
psychic      53
rock         45
steel        24
water       114
dtype: int64

if you want to collect multiple statistics together you can aggregate them with .agg()
.agg() can be used on it's own or with .groupby()

In [96]:
pokemon_groupby_agg = pokemon_groupby.agg(['max', 'min', 'median'])
pokemon_groupby_agg

/tmp/ipykernel_25061/1823518540.py:1: FutureWarning: ['pokemon'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.
  pokemon_groupby_agg = pokemon_groupby.agg(['max', 'min', 'median'])


deck_no             attack            defense            sp_attack  \
             max  min median    max min median     max min median       max   
type                                                                          
bug          795   10  407.0    185  10   65.0     230  30   60.0       145   
dark         799  197  559.0    150  50   88.0     125  30   70.0       140   
dragon       784  147  610.0    180  50  100.0     130  35   90.0       180   
electric     796   25  403.0    123  30   65.0     115  15   60.0       173   
fairy        764   35  669.5    131  20   58.5      95  28   67.5       131   
fighting     766   56  447.5    145  35  100.0      95  30   66.0       140   
fire         776    4  357.0    160  30   79.5     140  37   61.5       159   
flying       715  641  714.0    100  30   70.0      80  35   80.0       110   
ghost        781   92  487.0    165  30   66.0     145  30   80.0       170   
grass        798    1  420.5    181  27   68.0     131  30   66.0       145   
ground       750   27  363.5    180  40   90.0     160  30   82.0       150   
ice          713  124  378.0    130  30   65.0     184  15   70.0       130   
normal       780   16  335.0    160   5   75.0     126   5   60.0       135   
poison       758   23  316.5    106  43   68.5     152  35   66.0       111   
psychic      800   63  480.0    165  20   55.0     131  15   65.0       194   
rock         793   74  409.0    165  40   89.0     200  40   95.0       160   
steel        801  208  436.5    150  24   90.0     230  50  115.0       150   
water        788    7  320.5    155  10   70.0     180  20   70.0       180   

          ... capture_rate total_bs             gen            legendary      \
          ...       median      max  min median max min median       max min   
type      ...                                                                  
bug       ...         90.0      600  194  392.5   7   1    4.0         1   0   
dark      ...         45.0      680  220  480.0   7   2    5.0         1   0   
dragon    ...         45.0      780  300  540.0   7   1    5.0         1   0   
electric  ...         75.0      610  205  435.0   7   1    4.0         1   0   
fairy     ...        130.0      680  218  456.0   7   1    6.0         1   0   
fighting  ...         75.0      625  210  455.0   7   1    4.0         0   0   
fire      ...         45.0      680  250  473.0   7   1    3.5         1   0   
flying    ...         45.0      580  245  535.0   6   5    6.0         1   0   
ghost     ...         90.0      680  275  474.0   7   1    4.0         1   0   
grass     ...         75.0      630  180  409.0   7   1    4.0         1   0   
ground    ...         90.0      770  265  427.5   7   1    3.0         1   0   
ice       ...         60.0      580  250  455.0   6   1    3.0         1   0   
normal    ...         90.0      720  190  415.0   7   1    3.0         1   0   
poison    ...        105.0      535  245  406.5   7   1    3.0         0   0   
psychic   ...         50.0      780  198  470.0   7   1    4.0         1   0   
rock      ...         45.0      700  280  460.0   7   1    4.0         1   0   
steel     ...         37.5      700  300  490.0   7   2    4.0         1   0   
water     ...         60.0      770  200  440.0   7   1    3.0         1   0   

                 
         median  
type             
bug         0.0  
dark        0.0  
dragon      0.0  
electric    0.0  
fairy       0.0  
fighting    0.0  
fire        0.0  
flying      0.0  
ghost       0.0  
grass       0.0  
ground      0.0  
ice         0.0  
normal      0.0  
poison      0.0  
psychic     0.0  
rock        0.0  
steel       0.0  
water       0.0  

[18 rows x 30 columns]

You can ask for different statistics on different rows.


pokemon_agg_select = pokemon_groupby.agg({'attack':['max','min'],'defense':['max','min'], 'total_bs':['median']})

In [98]:
pokemon_agg_select = pokemon_groupby.agg({'attack':['max','min'],'defense':['max','min'], 'total_bs':['median']})
pokemon_agg_select

attack     defense     total_bs
            max min     max min   median
type                                    
bug         185  10     230  30    392.5
dark        150  50     125  30    480.0
dragon      180  50     130  35    540.0
electric    123  30     115  15    435.0
fairy       131  20      95  28    456.0
fighting    145  35      95  30    455.0
fire        160  30     140  37    473.0
flying      100  30      80  35    535.0
ghost       165  30     145  30    474.0
grass       181  27     131  30    409.0
ground      180  40     160  30    427.5
ice         130  30     184  15    455.0
normal      160   5     126   5    415.0
poison      106  43     152  35    406.5
psychic     165  20     131  15    470.0
rock        165  40     200  40    460.0
steel       150  24     230  50    490.0
water       155  10     180  20    440.0

missing section on math - conditional replacement - refer to module 2 for information

for Module 1 and 2

Module 3 and 4 combined below

Tidy Data!
Each row is a single observation
each variable is a single column
each value is a single cell

Pivoting transforms long dataframes into wider ones.
This is usefull if the dataframe has duplicate rows due to the data not being tidy.

.pivot()

takes the following arguements
index - uesed to make the new dataframes index
columns - used to select which column you want to transform
values - the column name where the values of the column you want to pivot exists

when using pivot we loose the other columns. 

Use .pivot_table() in most cases

also remember to .reset_index()

.rename_axis('') - will remove the index column

.pivot_table() takes averages of duplicate columns

.duplicated() will remove the duplicated data




In [104]:
pokemon_remove_duplicate=pokemon.duplicated(subset='total_bs', keep=False)
pokemon_remove_duplicate

0      True
1      True
2      True
3      True
4      True
       ... 
796    True
797    True
798    True
799    True
800    True
Length: 801, dtype: bool

In [102]:
pokemon[pokemon_remove_duplicate]

pokemon  deck_no  attack  defense  sp_attack  sp_defense  speed  \
0     Bulbasaur        1      49       49         65          65     45   
1       Ivysaur        2      62       63         80          80     60   
2      Venusaur        3     100      123        122         120     80   
3    Charmander        4      52       43         60          50     65   
4    Charmeleon        5      64       58         80          65     80   
..          ...      ...     ...      ...        ...         ...    ...   
796  Celesteela      797     101      103        107         101     61   
797     Kartana      798     181      131         59          31    109   
798    Guzzlord      799     101       53         97          53     43   
799    Necrozma      800     107      101        127          89     79   
800    Magearna      801      95      115        130         115     65   

     capture_rate  total_bs     type  gen  legendary  
0              45       318    grass    1          0  
1              45       405    grass    1          0  
2              45       625    grass    1          0  
3              45       309     fire    1          0  
4              45       405     fire    1          0  
..            ...       ...      ...  ...        ...  
796            25       570    steel    7          1  
797           255       570    grass    7          1  
798            15       570     dark    7          1  
799             3       600  psychic    7          1  
800             3       600    steel    7          1  

[801 rows x 12 columns]

bad example since it's a tidy data frame, but if there was data hiding in the 801 rows then this would help us identify it.

The code below pivoted the data so the name of the pokemon is the rows and the columns are now 'type' because we didn't put in a values arguement

In [108]:
pokemon_pivot=pokemon.pivot_table(index='pokemon',columns='type').reset_index()
pokemon_pivot

pokemon attack                                                    \
type               bug   dark dragon electric fairy fighting fire flying   
0     Abomasnow    NaN    NaN    NaN      NaN   NaN      NaN  NaN    NaN   
1          Abra    NaN    NaN    NaN      NaN   NaN      NaN  NaN    NaN   
2         Absol    NaN  150.0    NaN      NaN   NaN      NaN  NaN    NaN   
3      Accelgor   70.0    NaN    NaN      NaN   NaN      NaN  NaN    NaN   
4     Aegislash    NaN    NaN    NaN      NaN   NaN      NaN  NaN    NaN   
..          ...    ...    ...    ...      ...   ...      ...  ...    ...   
795     Zoroark    NaN  105.0    NaN      NaN   NaN      NaN  NaN    NaN   
796       Zorua    NaN   65.0    NaN      NaN   NaN      NaN  NaN    NaN   
797       Zubat    NaN    NaN    NaN      NaN   NaN      NaN  NaN    NaN   
798    Zweilous    NaN   85.0    NaN      NaN   NaN      NaN  NaN    NaN   
799     Zygarde    NaN    NaN  100.0      NaN   NaN      NaN  NaN    NaN   

            ... total_bs                                                      \
type ghost  ...    ghost  grass ground ice normal poison psychic rock  steel   
0      NaN  ...      NaN  594.0    NaN NaN    NaN    NaN     NaN  NaN    NaN   
1      NaN  ...      NaN    NaN    NaN NaN    NaN    NaN   310.0  NaN    NaN   
2      NaN  ...      NaN    NaN    NaN NaN    NaN    NaN     NaN  NaN    NaN   
3      NaN  ...      NaN    NaN    NaN NaN    NaN    NaN     NaN  NaN    NaN   
4      NaN  ...      NaN    NaN    NaN NaN    NaN    NaN     NaN  NaN  520.0   
..     ...  ...      ...    ...    ...  ..    ...    ...     ...  ...    ...   
795    NaN  ...      NaN    NaN    NaN NaN    NaN    NaN     NaN  NaN    NaN   
796    NaN  ...      NaN    NaN    NaN NaN    NaN    NaN     NaN  NaN    NaN   
797    NaN  ...      NaN    NaN    NaN NaN    NaN  245.0     NaN  NaN    NaN   
798    NaN  ...      NaN    NaN    NaN NaN    NaN    NaN     NaN  NaN    NaN   
799    NaN  ...      NaN    NaN    NaN NaN    NaN    NaN     NaN  NaN    NaN   

            
type water  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN  
..     ...  
795    NaN  
796    NaN  
797    NaN  
798    NaN  
799    NaN  

[800 rows x 181 columns]

In [111]:
pokemon_pivot.columns

Index(['pokemon', 'bug', 'dark', 'dragon', 'electric', 'fairy', 'fighting',
       'fire', 'flying', 'ghost', 'grass', 'ground', 'ice', 'normal', 'poison',
       'psychic', 'rock', 'steel', 'water'],
      dtype='object', name='type')

In [110]:
pokemon_pivot=pokemon.pivot_table(index='pokemon',columns='type',values='deck_no').reset_index()
pokemon_pivot

type    pokemon    bug   dark  dragon  electric  fairy  fighting  fire  \
0     Abomasnow    NaN    NaN     NaN       NaN    NaN       NaN   NaN   
1          Abra    NaN    NaN     NaN       NaN    NaN       NaN   NaN   
2         Absol    NaN  359.0     NaN       NaN    NaN       NaN   NaN   
3      Accelgor  617.0    NaN     NaN       NaN    NaN       NaN   NaN   
4     Aegislash    NaN    NaN     NaN       NaN    NaN       NaN   NaN   
..          ...    ...    ...     ...       ...    ...       ...   ...   
795     Zoroark    NaN  571.0     NaN       NaN    NaN       NaN   NaN   
796       Zorua    NaN  570.0     NaN       NaN    NaN       NaN   NaN   
797       Zubat    NaN    NaN     NaN       NaN    NaN       NaN   NaN   
798    Zweilous    NaN  634.0     NaN       NaN    NaN       NaN   NaN   
799     Zygarde    NaN    NaN   718.0       NaN    NaN       NaN   NaN   

type  flying  ghost  grass  ground  ice  normal  poison  psychic  rock  steel  \
0        NaN    NaN  460.0     NaN  NaN     NaN     NaN      NaN   NaN    NaN   
1        NaN    NaN    NaN     NaN  NaN     NaN     NaN     63.0   NaN    NaN   
2        NaN    NaN    NaN     NaN  NaN     NaN     NaN      NaN   NaN    NaN   
3        NaN    NaN    NaN     NaN  NaN     NaN     NaN      NaN   NaN    NaN   
4        NaN    NaN    NaN     NaN  NaN     NaN     NaN      NaN   NaN  681.0   
..       ...    ...    ...     ...  ...     ...     ...      ...   ...    ...   
795      NaN    NaN    NaN     NaN  NaN     NaN     NaN      NaN   NaN    NaN   
796      NaN    NaN    NaN     NaN  NaN     NaN     NaN      NaN   NaN    NaN   
797      NaN    NaN    NaN     NaN  NaN     NaN    41.0      NaN   NaN    NaN   
798      NaN    NaN    NaN     NaN  NaN     NaN     NaN      NaN   NaN    NaN   
799      NaN    NaN    NaN     NaN  NaN     NaN     NaN      NaN   NaN    NaN   

type  water  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN  
..      ...  
795     NaN  
796     NaN  
797     NaN  
798     NaN  
799     NaN  

[800 rows x 19 columns]

.melt() Arugements:

id_vars = columns you want to keep
value_vars = columns you want ot melt
var_nam e= name of the new column
value_name = measurements for the new column are housed

This took the data set and melted all the bug and water columns together.



In [122]:
melted_pokemon=(pokemon_pivot.melt(id_vars='pokemon',
                            value_vars=['bug','water'], 
                            var_name='type',
                            value_name=''))
melted_pokemon

pokemon   type       
0     Abomasnow    bug    NaN
1          Abra    bug    NaN
2         Absol    bug    NaN
3      Accelgor    bug  617.0
4     Aegislash    bug    NaN
...         ...    ...    ...
1595    Zoroark  water    NaN
1596      Zorua  water    NaN
1597      Zubat  water    NaN
1598   Zweilous  water    NaN
1599    Zygarde  water    NaN

[1600 rows x 3 columns]

Joining Data frames

pd.concat() - a forcefull way of joing dataframes across rows or columns

.merge() - combinds data on common columns or indices